flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -537,13 +545,30 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
 			if (backlog >= 0) {
 				onSenderBacklog(backlog);
 			}
+
+			if (checkpointBarrier != null && inputGate.listener != null) {
+				inputGate.listener.notifyBarrierReceived(checkpointBarrier, channelIndex);
+			}
 		} finally {
 			if (recycleBuffer) {
 				buffer.recycleBuffer();
 			}
 		}
 	}
 
+	@Nullable
+	private CheckpointBarrier parseCheckpointBarrier(Buffer buffer) throws IOException {
```

Comment:
Would that be a case for `Optional`? Or do we not use Optional for performance criticial path?

Reply:
It is a buffer level operation, so I guess it is not performance sensitive.  But is is also not a API/public methods which would be suggested to use `Optional` generally by https://flink.apache.org/contributing/code-style-and-quality-java.html#java-optional.

Actually I did not consider much in PoC.